File: lib/forward-checking.ts
import MapColoringGraph from "./map-coloring-graph";
import {
  average,
  getNeighbours,
  getUnassigned,
  isAssignmentValid,
  median,
  standardDeviation,
} from "./utils";

export default class ForwardChecking {
  mapColoringGraph: MapColoringGraph;
  mapColoringGraphs: MapColoringGraph[];
  assignments: number[];
  exampleAssignments: number[];
  averageRunTime: number;
  standardDeviation: number;
  medianRunTime: number;
  k: number;
  backtracks: number;

  constructor(mcgs: MapColoringGraph[], k: number) {
    this.mapColoringGraphs = mcgs;
    this.mapColoringGraph = mcgs[0];
    this.k = k;
    this.averageRunTime = 0;
    this.standardDeviation = 0;
    this.medianRunTime = 0;
    this.assignments = [];
    this.exampleAssignments = [];
    this.backtracks = 0;
  }

  run() {
    const times: number[] = [];
    const backtracksList: number[] = [];

    for (let i = 0; i < this.mapColoringGraphs.length; i++) {
      this.mapColoringGraph = this.mapColoringGraphs[i];

      this.assignments = [];
      this.backtracks = 0;

      const start = performance.now();
      this.backtrack();
      const end = performance.now();

      times.push(end - start);
      backtracksList.push(this.backtracks);

      if (i === 0) {
        this.exampleAssignments = this.assignments;
      }
    }

    this.averageRunTime = average(times);
    this.standardDeviation = standardDeviation(times);
    this.medianRunTime = median(times);
    this.backtracks = average(backtracksList);
  }

  private backtrack() {
    const domains: number[][] = [];
    const assignments: number[] = [];

    for (let i = 0; i < this.mapColoringGraph.n; i++) {
      domains.push(Array.from(Array(this.k).keys()));
      assignments.push(-1);
    }

    const res = this.backtrackWithForwardChecking(assignments, domains);
    if (res) {
      // Backtracking with Forward Checking has found a valid coloring
      this.assignments = assignments;
    }
  }

  private backtrackWithForwardChecking(
    assignments: number[],
    domains: number[][]
  ) {
    // Check if all map points have been assigned a color, if so a valid coloring has been found
    if (assignments.every((assignment) => assignment !== -1)) {
      return true;
    }

    // Get the index of the unassigned map point with the smallest domain (Minimum Remaining Values)
    const unassigned = getUnassigned(assignments, domains);

    for (const value of domains[unassigned]) {
      const tempDomains = [...domains.map((domain) => [...domain])];
      assignments[unassigned] = value;
      domains[unassigned] = [value];

      // Get the neighbors of the current unassigned point
      const neighbours = getNeighbours(
        this.mapColoringGraph.points,
        this.mapColoringGraph.lines,
        this.mapColoringGraph.points[unassigned]
      );

      for (let neighbour of neighbours) {
        // Skip already assigned neighbors
        if (assignments[neighbour] !== -1) {
          continue;
        }

        // Filter the domains array for each neighbor to remove colors that violate the constraint
        domains[neighbour] = domains[neighbour].filter((domainValue) => {
          return isAssignmentValid(
            this.mapColoringGraph.points,
            this.mapColoringGraph.lines,
            assignments,
            neighbour,
            domainValue
          );
        });
      }

      if (domains.some((domain) => domain.length === 0)) {
        assignments[unassigned] = -1;
        domains = tempDomains;
        continue;
      }

      // Recursively call the method with updated assignments and domains
      if (this.backtrackWithForwardChecking(assignments, domains)) {
        return true;
      }

      // Reset the assignment of the current unassigned point and restore the previous domains
      assignments[unassigned] = -1;
      domains = tempDomains;
    }

    // No valid coloring found, backtrack
    this.backtracks++;
    return false;
  }

  toString() {
    return `Average run time: ${this.averageRunTime} ms\nStandard deviation: ${this.standardDeviation} ms\nMedian run time: ${this.medianRunTime} ms`;
  }
}

File: py/lib/forward_checking.py
import time
import copy
from typing import List
from lib.map_coloring_graph import MapColoringGraph 
from lib.utils import average, median, standard_deviation  

class ForwardChecking:
    def __init__(self, mcg: MapColoringGraph, k: int,  iterations: int):
        self.map_coloring_graph = mcg
        self.k = k
        self.average_run_time = 0
        self.standard_deviation = 0
        self.median_run_time = 0
        self.iterations = iterations
        self.assignments = []

    def run(self):
        times = []

        for _ in range(self.iterations):
            start = time.perf_counter()
            self.backtrack()
            end = time.perf_counter()
            times.append(end - start)

        self.average_run_time = average(times)
        self.standard_deviation = standard_deviation(times)
        self.median_run_time = median(times)

        self.plot_solution()

    def backtrack(self):
        domains = []
        assignments = []

        for _ in range(self.map_coloring_graph.n):
            domains.append(list(range(self.k)))
            assignments.append(-1)

        res = self.backtrack_with_forward_checking(assignments, domains)
        if res:
            self.assignments = assignments
            # print(assignments)
            # print("Solution found")
        # else:
            # print("No solution found")

    def backtrack_with_forward_checking(self, assignments: List[int], domains: List[List[int]]):
        if all(assignment != -1 for assignment in assignments):
            return True

        unassigned = self.get_unassigned(assignments)

        # print(assignments, domains)

        for value in domains[unassigned]:
            temp_domains = [list(domain) for domain in domains]
            assignments[unassigned] = value
            domains[unassigned] = [value]

            for i in range(len(assignments)):
                if assignments[i] != -1 or i == unassigned:
                    continue

                domains[i] = [domain_value for domain_value in domains[i] if
                              self.is_assignment_valid(assignments, i, domain_value)]

            if self.backtrack_with_forward_checking(assignments, domains):
                return True

            assignments[unassigned] = -1
            domains = temp_domains

        return False

    def get_unassigned(self, assignments: List[int]):
        for i, assignment in enumerate(assignments):
            if assignment == -1:
                return i
        return -1

    def is_assignment_valid(self, assignments: List[int], unassigned: int, value: int):
        for i in range(len(assignments)):
            if assignments[i] == -1 or i == unassigned:
                continue

            if any(line.pointA == self.map_coloring_graph.points[i] and line.pointB == self.map_coloring_graph.points[unassigned] or
                   line.pointA == self.map_coloring_graph.points[unassigned] and line.pointB == self.map_coloring_graph.points[i]
                   for line in self.map_coloring_graph.lines) and assignments[i] == value:
                return False

        return True

    def plot_solution(self):
        colors = ['red', 'blue', 'green', 'yellow']
        color_assignments = []

        for i in range(len(self.assignments)):
            color_assignments.append(colors[self.assignments[i]])

        print(color_assignments)

        self.map_coloring_graph.plot_grid("output/map_coloring_" + str(self.map_coloring_graph.n) + "_" + str(self.k) + "_forward_checking.png", color_assignments)


    def __str__(self):
        return f"Average run time: {self.average_run_time} s\nStandard deviation: {self.standard_deviation} s\nMedian run time: {self.median_run_time} s"
